### PR TITLE
New example program for printing supported features

### PR DESCRIPTION
**Pull request by baierd**
This pull request is for issue #192 and #194 as it seems to be a small problem that was found by developing the example program.

As i found some inconsistencies with MathSAT those will be included here.
I've added a null check in the `UnsatCoreWithAssumptions()` method of MathSAT5 as it did not act like the other solvers.
Furthermore i added a test for this behaviour.
The method `UnsatCoreWithAssumptions()` itself seems to have a problem as seen in #194 and will be checked as well.

A first iteration of the program is included.
Main problem at the moment is that the output of the example program in the script and therefore the terminal puts line breaks on its own depending on the width of the terminal.
@kfriedberger is there a easy method to fix this?

### CODE REVIEW

**Comment by kfriedberger** [comment]
The table is quite nice and clearly written, I like it. It is just very broad.
A possible solution would be to use multiple lines per cell, but that might be quite tricky to implement in a simple and understandable way.
And perhaps, we should aim for a more object-oriented approach instead of using lists of lists of Strings.

File: src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+
+  /**
+   * Checks the infoArray for the length of the given column.
+   *
+   * @param columnToCheck column of the infoArray you want the length for.
+   * @return length of String in column, but at least 8 (max length of the descriptor Strings).
+   */
+  private int lineLength(List<List<String>> infoArray, int columnToCheck) {
+    if (columnToCheck > 3 || columnToCheck < 0) {
+      return 8;
```
Comment (kfriedberger) [inline src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java:300]: This looks like a constant MIN_CELL_WIDTH.

File: src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+    }
+    int length = 8;
```
Comment (kfriedberger) [inline src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java:302]: MIN_CELL_WIDTH?

File: src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+    List<String> features = new ArrayList<>();
+
+    // Optimization: Will throw UnsupportedOperationException in creation of prover if not
+    // available.
+    try (OptimizationProverEnvironment prover =
+        context.newOptimizationProverEnvironment(ProverOptions.GENERATE_MODELS)) {
+      if (prover != null) {
+        features.add("Optimization");
+      }
+    } catch (UnsupportedOperationException e) {
```
Comment (kfriedberger) [inline src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java:175]: add a comment, why the exception can be ignored.

File: src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+    }
+
+    // Interpolation: Will throw UnsupportedOperationException in creation of prover if not
+    // available.
+    try (InterpolatingProverEnvironment<?> prover =
+        context.newProverEnvironmentWithInterpolation(ProverOptions.GENERATE_MODELS)) {
+      if (prover != null) {
+        features.add("Interpolation");
+      }
+    } catch (UnsupportedOperationException e) {
```
Comment (kfriedberger) [inline src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java:185]: add a comment, why the exception can be ignored.

File: src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+
+    // UnsatCore: throws UnsupportedOperationException if not available.
+    // TODO: check behavior with null argument properly
+    try (ProverEnvironment prover =
+        context.newProverEnvironment(ProverOptions.GENERATE_UNSAT_CORE)) {
+      if (prover.getUnsatCore() != null) {
+        features.add("UnsatCore");
+      }
+    } catch (UnsupportedOperationException e) {
+    } catch (Exception e) {
```
Comment (kfriedberger) [inline src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java:196]: Why is this needed?

File: src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+      features.add("UnsatCore");
+    }
+
+    // UnsatCoreOverAssumptions: throws NullPointerException if available.
+    try (ProverEnvironment prover =
+        context.newProverEnvironment(GENERATE_UNSAT_CORE_OVER_ASSUMPTIONS)) {
+      prover.unsatCoreOverAssumptions(null);
+    } catch (UnsupportedOperationException e) {
+    } catch (NullPointerException e) {
```
Comment (kfriedberger) [inline src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java:205]: add a comment, why the exception can be ignored.

File: src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+                  public List<List<BooleanFormula>> getResult() {
+                    return models;
+                  }
+                },
+                ImmutableList.of())
+            != null) {
+          features.add("AllSAT");
+        }
+      } catch (UnsupportedOperationException e) {
+      }
```
Comment (kfriedberger) [inline src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java:233]: add a comment, why the exception can be ignored.

File: src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+    }
+
+    // We don't care about the return value, just that it doesn't throw an
+    // UnsupportedOperationException.
+    try (ProverEnvironment prover = context.newProverEnvironment(ProverOptions.GENERATE_MODELS)) {
+      try {
+        prover.isUnsatWithAssumptions(ImmutableList.of());
+        features.add("Assumptions");
+      } catch (UnsupportedOperationException e) {
+      }
```
Comment (kfriedberger) [inline src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java:243]: add a comment, why the exception can be ignored.

File: src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+    }
+
+    // UnsatCore: throws UnsupportedOperationException if not available.
+    // TODO: check behavior with null argument properly
+    try (ProverEnvironment prover =
+        context.newProverEnvironment(ProverOptions.GENERATE_UNSAT_CORE)) {
+      if (prover.getUnsatCore() != null) {
+        features.add("UnsatCore");
+      }
+    } catch (UnsupportedOperationException e) {
```
Comment (kfriedberger) [inline src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java:195]: add a comment, why the exception can be ignored.

File: src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+   * 4.
+   *
+   * @param solver to check for information. Taken from Solvers enum only.
+   * @return String[] of length 4 of the solver you entered. Array content as String in the
+   *     following order: SolverName, SolverVersion, SolverTheories, SolverFeatures. String[] length
+   *     0 if invalid solver.
+   */
+  private List<String> getSolverInformation(Solvers solver)
+      throws SolverException, InterruptedException {
+    List<String> info = new ArrayList<>();
```
Comment (kfriedberger) [inline src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java:134]: An `info` object with named members would be nicer .... perhaps.

File: src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java (hunk 5 above)
Comment (baierd) [reply src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java:196]: Because i wanted minimal examples to check for `UnsupportedOperationException` and didn't yet check how the method `getUnsatCore()` behaves on an empty stack. This is subject to change.

File: src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java (hunk 10 above)
Comment (baierd) [reply src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java:134]: `info` object in a Class?
Could you explain further please?

Comment (baierd) [reply src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java:134]: > The table is quite nice and clearly written, I like it. It is just very broad.
> A possible solution would be to use multiple lines per cell, but that might be quite tricky to implement in a simple and understandable way.
Yes i know. But i know that its possible to "unlock" the terminal/shell so that you can scroll and that there are no forced line breaks, i just dont know how yet.

> And perhaps, we should aim for a more object-oriented approach instead of using lists of lists of Strings.
Done

**Comment by baierd** [comment]
I know its very broad and thats a problem in the terminal. But i know that its possible to stop the terminal from forcing line breaks somehow, i've seen it. Either i will find out or change the format.

File: src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java (hunk 10 above)
Comment (kfriedberger) [reply src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java:134]: The SolverInfo class would allow to split data retrieval and data dumping. Just an idea.
The following draft might help:
```
/** data class containing the info for the solver and can print it */
class SolverInfo {
  - Solvers solver;
  - String version;
  - List<String> supportedTheories;
  - List<String> features;
}
/** helper class for wrapping the contents of a row in the table  */
class Row {
  - List<String> lines;
  - int minNumOfLines;
  - List<Integer> columnWidths; // needed for producing a new empty line with seperators "|   |   |   |"
  + appendCell(List<String> lines); // append cell-lines to lines. Add new line, if number of cell lines is too large.   
  + toString(); // dump nice row, include terminal line as last line: "|---|---|---|"
}
+ List<String> cellToRow(String cellContent, int exactLengthOfCell); // returns a list of lines, formatted to match exact length or less space, does no contain seperator "|"
```

**Comment by kfriedberger** [comment]
Some terminal can disable line breaks, some do not.
We dot want to rely on terminal-specific code, because that might then be broken on windows... :-)

For me, the table would also be ok, as it is now, and it helps a lot as documentation.

**Comment by baierd** [comment]
Do you know a good width that works in most cases?
Something similar as the other examples?

**Comment by kfriedberger** [comment]
Lets try, how Github layouts the table on a normal monitor (table written in MarkDown, with ", " as seperator in theory and feature cell):

| Solver | Version | Theories | Features |
| --- | --- | --- | --- |
| MATHSAT5 | MathSAT5 version 5.6.3 (973d615f065f) (May 12 2020 08:45:26, gmp 6.1.0, gcc 4.8.5, 64-bit, reentrant) | Integer, Bitvector, Rational, Float, Array, UF | Optimization, Interpolation, UnsatCore, UnsatCore /w Assumption, AllSAT, Assumptions |
| SMTINTERPOL | SMTInterpol 2.5-604-g71e72f93 | Integer, Rational, Array, UF | Interpolation, UnsatCore, UnsatCore /w Assumption, AllSAT, Assumptions |
| Z3 | Z3 4.8.8.0 | Integer, Bitvector, Rational, Float, Array, Quantifier, UF | Optimization, UnsatCore, UnsatCore /w Assumption, AllSAT, Assumptions |
| PRINCESS | Princess (2020-05-01 (assertionless)) | Integer, Bitvector, Array, Quantifier, UF | Interpolation, UnsatCore, AllSAT, Assumptions |
| BOOLECTOR   | Boolector 3.2.1 | Bitvector, Array, Quantifier, UF | AllSAT, Assumptions |
| CVC4 | CVC4 1.8-prerelease | Integer, Bitvector, Rational, Float, Array, UF, Seperation-Logic | UnsatCore, AllSAT, Assumptions |

Overall width for columns:
| 15 (full name) | 40 (split at space) | 25 (split at space) | 25 (split at space) |
| --- | --- | --- | --- |

Would this be possible?

**Comment by baierd** [comment]
Looks good.
I'll give it a try.

**Comment by baierd** [comment]
This table should be fine now.
Code needs to be cleaned up though.

**Comment by kfriedberger** [comment]
Thanks for the nice table and formatting code.
I applied some cleanup. It looks fine now.

**Comment by kfriedberger** [comment]
@baierd do you have any more ideas for improvement or can I merge it?

**Comment by baierd** [comment]
If its ok i'd like to try to check if AllSat is overwritten in the solver and check the behaviour of UnsatCore on an empty stack.

**Comment by kfriedberger** [comment]
When merging the master into this branch, Yices2 fails without a clear reason. We need to fix this in the Yices2-bindings.

**Comment by baierd** [comment]
OK i'll take a look.

File: src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java (hunk 5 above)
Comment (baierd) [reply src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java]: `UnsatCore()` throws different Exceptions based on the solver used (native Exceptions). I added a comment.

**Comment by baierd** [comment]
I still have a problem with `AllSat()`. If i use reflections to get the method or class i run into basicly the same problem. The class of a prover initialized with `GENERATE_ALL_SAT` is not uniform. For example SMTInterpol has a `ProverWithAssumptionsWrapper` and overrides `AllSat()` while CVC4 returns the same class and does not support it.
You can't check for thrown Exceptions either as the `UnsupportedOperationException` is not in the method declaration.

As the current check for AllSat doesn't work and you said its not important, i would simply remove it.

**Comment by kfriedberger** [comment]
Ok, just remove AllSAT from the table.

**Comment by baierd** [comment]
Thats maybe a stupid question, but shouldn't a prover thats initialized with `GENERATE_UNSAT_CORE_OVER_ASSUMPTIONS` be able to perform UnsatCore, UnsatCore /w Assumptions and also Assumptions (without UnsatCore). 
If yes we could combine that testcase into 1 prover object.

**Comment by kfriedberger** [comment]
Yes, `UNSAT_CORE_OVER_ASSUMPTIONS` implies the other features. But it is not the other way around, e.g., `UNSAT_CORE` might be supported, but the rest is not.

**Comment by baierd** [comment]
Yices2 works now. I had to reorder the way the features are checked.

Yices had a problem with the config it uses to create a new prover (yices2 context). Once you use `UnsatCore `once you may not use DPLL or push/pop again. If you do it ends in an `IllegalArgumentException `(native code returns `CTX_INVALID_PARAMETER_VALUE` for error in changing the config). 
Is UnsatCore considered non-linear arithmetics? Because in that case Yices2 defaults to MCSat which does not support push/pop. If not i don't know why.

**Comment by kfriedberger** [comment]
Is the reordering now hiding a bug in Yices2?
Is creating and deleting a prover in yices not cleaning up properly and leaving behind traces/options that are unchangable?

UnsatCores have nothing to do with non-linear arithmetics, they are an orthogonal feature.
We require push/pop, and that should be available for other backend solvers.

**Comment by baierd** [comment]
Possibly.
I tried some things:
-creating a static config that is not deleted with push/pop works fine with this program, but fails in the tests severly!
  (Probably because non-linear arithmetics needs another config acording to the API. But i don't know why our tests don't fail 
  then as we use the config with DPLL and push/pop every time. Maybe it simply overrides the settings after using it?)
-calling UnsatCoreWithAssumptions before UnsatCore is fine, the other way fails. Same with Assumptions and UnsatCore.

**Comment by kfriedberger** [comment]
Looks good now. Anything remaining for this branch?

**Comment by baierd** [comment]
I think this is fine. You can merge.
